src/index.ts

```diff
@@ -74,253 +74,272 @@ const TYPED_MESSAGE_SCHEMA = {
+): string {
+  let result = '';
+  let deps = findTypeDependencies(primaryType, types).filter(
+    (dep) => dep !== primaryType,
+  );
+  deps = [primaryType].concat(deps.sort());
+  for (const type of deps) {
+    const children = types[type];
+    if (!children) {
+      throw new Error(`No type definition specified: ${type}`);
     }
-    return result;
-  },
+    result += `${type}(${types[type]
+      .map(({ name, type: t }) => `${t} ${name}`)
+      .join(',')})`;
+  }
+  return result;
+}
 
-  /**
-   * Finds all types within a type definition object
-   *
-   * @param {string} primaryType - Root type
-   * @param {Object} types - Type definitions
-   * @param {Array} results - current set of accumulated types
-   * @returns {Array} - Set of all types found in the type definition
-   */
-  findTypeDependencies(
-    primaryType: string,
-    types: Record<string, MessageTypeProperty[]>,
-    results: string[] = [],
-  ): string[] {
-    [primaryType] = primaryType.match(/^\w*/u);
-    if (results.includes(primaryType) || types[primaryType] === undefined) {
-      return results;
-    }
-    results.push(primaryType);
-    for (const field of types[primaryType]) {
-      for (const dep of this.findTypeDependencies(field.type, types, results)) {
-        !results.includes(dep) && results.push(dep);
-      }
-    }
+/**
+ * Finds all types within a type definition object
+ *
+ * @param {string} primaryType - Root type
+ * @param {Object} types - Type definitions
+ * @param {Array} results - current set of accumulated types
+ * @returns {Array} - Set of all types found in the type definition
+ */
+function findTypeDependencies(
+  primaryType: string,
+  types: Record<string, MessageTypeProperty[]>,
+  results: string[] = [],
+): string[] {
+  [primaryType] = primaryType.match(/^\w*/u);
+  if (results.includes(primaryType) || types[primaryType] === undefined) {
     return results;
-  },
+  }
+  results.push(primaryType);
+  for (const field of types[primaryType]) {
+    for (const dep of findTypeDependencies(field.type, types, results)) {
+      !results.includes(dep) && results.push(dep);
+    }
+  }
+  return results;
+}
 
-  /**
-   * Hashes an object
-   *
-   * @param {string} primaryType - Root type
-   * @param {Object} data - Object to hash
-   * @param {Object} types - Type definitions
-   * @returns {Buffer} - Hash of an object
-   */
-  hashStruct(
-    primaryType: string,
-    data: Record<string, unknown>,
-    types: Record<string, unknown>,
-    useV4 = true,
-  ): Buffer {
-    return ethUtil.keccak(this.encodeData(primaryType, data, types, useV4));
-  },
+/**
+ * Hashes an object
+ *
+ * @param {string} primaryType - Root type
+ * @param {Object} data - Object to hash
+ * @param {Object} types - Type definitions
+ * @returns {Buffer} - Hash of an object
+ */
+function hashStruct(
+  primaryType: string,
+  data: Record<string, unknown>,
+  types: Record<string, MessageTypeProperty[]>,
```

Comment:
Here is the parameter type adjustment that was required to eliminate type errors. Clearly this is required because it's passed straight to `encodeData`, which itself required the `types` parameter to be of this type.